Add output table to the simulation result

anisometropie opened the pull request:
closes #7757 
closes #7758

Yohh commented on front/src/applications/operationalStudies/views/v2/SimulationResultsV2.tsx:
```diff
@@ -18,6 +18,9 @@ import SpeedSpaceChartV2 from 'modules/simulationResult/components/SpeedSpaceCha
 import TimeButtons from 'modules/simulationResult/components/TimeButtons';
 import TrainDetailsV2 from 'modules/simulationResult/components/TrainDetailsV2';
 import type { PositionScaleDomain, TimeScaleDomain } from 'modules/simulationResult/types';
+import TimesStops from 'modules/timesStops/TimesStops';
```
unused

```suggestion

```

anisometropie replied:
done

Yohh commented on front/src/applications/operationalStudies/views/v2/SimulationResultsV2.tsx:
```diff
@@ -18,6 +18,9 @@ import SpeedSpaceChartV2 from 'modules/simulationResult/components/SpeedSpaceCha
 import type { PositionScaleDomain, TimeScaleDomain } from 'modules/simulationResult/types';
+import TimesStops from 'modules/timesStops/TimesStops';
+import TimesStopsOutput from 'modules/timesStops/TimesStopsOutput';
+import { TableType } from 'modules/timesStops/types';
```
unused
```suggestion

```

anisometropie replied:
done

Yohh commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -58,10 +54,10 @@ const createDeleteViaButton = ({
 
-const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsProps) => {
+const TimesStops = ({ allWaypoints, pathSteps = [], startTime, tableType }: TimesStopsProps) => {
+  console.log('allWaypoints', allWaypoints);
```
```suggestion

```

anisometropie replied:
done

Yohh reviewed (commented):
needs a margin between the table and the slopes compoent

![image](https://github.com/OpenRailAssociation/osrd/assets/20640252/7a4837fe-039c-4847-b7ca-4204ef142a9d)

anisometropie commented:
> needs a margin between the table and the slopes compoent
> 
> ![image](https://private-user-images.githubusercontent.com/20640252/343709293-7a4837fe-039c-4847-b7ca-4204ef142a9d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0ODYyMjAsIm5iZiI6MTcxOTQ4NTkyMCwicGF0aCI6Ii8yMDY0MDI1Mi8zNDM3MDkyOTMtN2E0ODM3ZmUtMDM5Yy00ODQ3LWI3Y2EtNDIwNGVmMTQyYTlkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI3VDEwNTg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwOTY1MDM0ZDlmOTk1ODM3ZWRmYTU5ODRjMTQ1MjZmOWM2OTQwNjZiZDhkMDk1NThkMzNiY2VhNDE2NWIwMzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Zeg-XC7xYwPL77c1QZ5pA8lsxMDu0UWAywqkfanC8Xk)

fixed

SarahBellaha commented on front/public/locales/fr/timesStops.json:
```diff
@@ -1,10 +1,15 @@
   "theoreticalMarginPlaceholder": "% ou min/100km",
+  "calculatedArrivalTime": "Arrivée calculée",
+  "calculatedDepartureTime": "Départ calculé",
+  "realMargin": "Marge réeelle",
```
```suggestion
  "realMargin": "Marge réelle",
```

anisometropie replied:
done

theocrsb commented on front/public/locales/en/timesStops.json:
Keys need to be sorted alphabetically

anisometropie replied:
Done.

theocrsb commented on front/public/locales/fr/timesStops.json:
same here

anisometropie replied:
done

SarahBellaha commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,131 @@
+      op?.position * 1000,
+      'times'
+    );
+    const realMargin = (100 * (exactProvisionalTime - exactBaseTime)) / exactBaseTime;
```
When `exactBaseTime` === 0, it can return NaN / Infinity.
Maybe add this ?

```suggestion
    const realMargin = (100 * (exactProvisionalTime - exactBaseTime)) / exactBaseTime || 0;
```

Yohh commented on front/src/applications/operationalStudies/views/v2/ManageTrainScheduleV2.tsx:
```diff
@@ -18,6 +18,7 @@ import RollingStock2Img from 'modules/rollingStock/components/RollingStock2Img';
 import { RollingStockSelector } from 'modules/rollingStock/components/RollingStockSelector';
 import { useStoreDataForRollingStockSelector } from 'modules/rollingStock/components/RollingStockSelector/useStoreDataForRollingStockSelector';
 import TimesStops from 'modules/timesStops/TimesStops';
+import { TableType } from 'modules/timesStops/types';
```
add type prefix

anisometropie replied:
it’s an enum so I need to use it at runtime

anisometropie replied:
I reviewed all the imports, I think the other spots were also for this enum, I think the rest of imports are ok

Yohh commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -68,21 +70,17 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
+    const suggestedOPs = formatSuggestedViasToRowVias(allWaypoints, pathSteps, t, startTime);
     setTimesStopsSteps(suggestedOPs);
-  }, [t, pathProperties.allWaypoints, startTime]);
+  }, [t, allWaypoints, startTime]);
```
is `t` really necessary?

Yohh commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,235 @@
+      { skip: !selectedTrainId }
+    );
+
+  const tsScheduleByAt = keyBy(selectedTrainSchedule.schedule, 'at');
```
everywhere you create a const that starts with `ts`, can't you juste remove this start? I think `scheduleByAt` or `pathPointsWithOpPointIndices` or even `pathByUic`.. is enough

anisometropie replied:
it was too indicate that it’s data derived from train schedule. But yeah not really necessary.
It’s tough to find good names

Yohh reviewed (changes requested):
nice, I spotted many times where you forget the type prefix, I just commented one but you have the same issue in `TimeStopsColumns.tsx` and `TimesStops.tsx`

Yohh commented:
in SimulationResultsV2.tsx, still this block that seems useless

```js
      {/* SIMULATION: TIMELINE — TEMPORARILY DISABLED
      {simulation.trains.length && (
      {simulation.trains.length && (
        <TimeLine
        <TimeLine
          timeScaleDomain={timeScaleDomain}
          timeScaleDomain={timeScaleDomain}
          selectedTrainId={selectedTrain?.id || simulation.trains[0].id}
          selectedTrainId={selectedTrain?.id || simulation.trains[0].id}
          trains={simulation.trains as SimulationReport[]}
          trains={simulation.trains as SimulationReport[]}
          onTimeScaleDomainChange={setTimeScaleDomain}
          onTimeScaleDomainChange={setTimeScaleDomain}
        />
        />
      )}
      */}

kmer2016 commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,232 @@
+import { TableType } from './types';
+import { findNextScheduledPoint } from './utils';
+
+const getDigits = (u: string | undefined) => (u === 's' || u === '%' ? 0 : 1);
+function formatDigitsAndUnit(value: string | undefined, unit?: string) {
+  if (value === undefined || value === 'none') {
+    return '';
+  }
+  if (marginRegExValidation.test(value)) {
+    const [, v, , u] = value.match(marginRegExValidation) || [];
+    return `${stripDecimalDigits(Number(v), getDigits(u))}${NO_BREAK_SPACE}${u}`;
+  }
+  return `${stripDecimalDigits(Number(value), getDigits(unit))}${NO_BREAK_SPACE}${unit}`;
+}
```
should be in utils.tsx file ?

anisometropie replied:
yes it’s moved

kmer2016 commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,232 @@
+  return `${stripDecimalDigits(Number(value), getDigits(unit))}${NO_BREAK_SPACE}${unit}`;
+}
+
+type TimesStopsOutputProps = {
+  train: TrainScheduleBase;
+  simulatedTrain: SimulationResponseSuccess;
+  pathProperties: PathPropertiesFormatted;
+  infraId: number;
+};
```
shoud be in types.ts file

anisometropie replied:
it’s the file’s component props

clarani replied:
I'm fine with leaving the props of the component in the component file too :+1:

anisometropie replied:
ok leaving it there!

kmer2016 commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,232 @@
+  selectedTrainSchedule: TrainScheduleResult,
+  pathId: string
+) => {
+  // debugger;
```
```suggestion
```

anisometropie replied:
done

kmer2016 commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,232 @@
+};
+
+let marginUnit = '%';
+const computeMargins = (
```
Move this function in utils too

anisometropie replied:
done

anisometropie commented:
> nice, I spotted many times where you forget the type prefix, I just commented one but you have the same issue in `TimeStopsColumns.tsx` and `TimesStops.tsx`

I reviewed all the imports, I think the other spots were also for this enum, I think the rest of imports are ok

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
Don't remove this in this PR please, otherwise you should also remove the component TimeLine, but I think it's not the purpose of this PR

anisometropie replied:
ok leaving it as is

clarani commented on front/src/applications/operationalStudies/views/v2/SimulationResultsV2.tsx:
```diff
@@ -128,17 +129,6 @@ const SimulationResultsV2 = ({
-        />
-      )}
-      */}
-
```
Same here

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimeStopsColumns.tsx:
```diff
@@ -62,31 +98,32 @@ export const useInputColumns = (pathProperties: ManageTrainSchedulePathPropertie
-  );
+      ...extraOutputColumns,
+    ] as Column<PathWaypointRow>[];
+  }, [allWaypoints.length]);
```
```suggestion
  }, [tableType, allWaypoints.length]);
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimeStopsColumns.tsx:
```diff
@@ -52,6 +87,7 @@ export const useInputColumns = (pathProperties: ManageTrainSchedulePathPropertie
           })
         ),
         title: `${t('stopTime')}`,
```
```suggestion
        title: t('stopTime'),
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimeStopsColumns.tsx:
```diff
@@ -62,31 +98,32 @@ export const useInputColumns = (pathProperties: ManageTrainSchedulePathPropertie
         ),
-        cellClassName: ({ rowData }) => cx({ invalidCell: !rowData.isMarginValid }),
+        cellClassName: ({ rowData }) =>
+          cx({ invalidCell: tableType === TableType.Input && !rowData.isMarginValid }),
```
```suggestion
          cx({ invalidCell: !isOutputTable && !rowData.isMarginValid }),
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimeStopsColumns.tsx:
```diff
@@ -20,10 +20,37 @@ const timeColumn: Partial<Column<string | null | undefined, string, string>> = {
+    const extraOutputColumns = (
+      isOutputTable
+        ? [
+            {
+              ...keyColumn('calculatedMargin', createTextColumn()),
+              title: t('realMargin'),
+              disabled: true,
+            },
```
Maybe we can create a small utils function to create a disabled Text column ?

anisometropie replied:
sure, good idea. it’s done

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -54,10 +48,18 @@ const createDeleteViaButton = ({
-const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsProps) => {
-  const { t } = useTranslation('timesStops');
+type TimesStopsProps = {
+  // review insight: we were passing the entire ManageTrainSchedulePathProperties but only using allWaypoints
```
Yes I agree, thanks for the cleaning :)
```suggestion
```

anisometropie replied:
removed the insight, who now has served its purpose :)

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -54,10 +48,18 @@ const createDeleteViaButton = ({
-  const { t } = useTranslation('timesStops');
+type TimesStopsProps = {
+  // review insight: we were passing the entire ManageTrainSchedulePathProperties but only using allWaypoints
+  allWaypoints: SuggestedOP[] | undefined;
```
```suggestion
  allWaypoints?: SuggestedOP[];
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -68,21 +70,17 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
 
-  const [timesStopsSteps, setTimesStopsSteps] = useState<PathWaypointColumn[]>([]);
+  // The rows of the table
+  const [timesStopsSteps, setTimesStopsSteps] = useState<PathWaypointRow[]>([]);
```
Maybe we can rename this const rows ?

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -99,41 +97,64 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
+      className="time-stops-datasheet"
       columns={columns}
       value={timesStopsSteps}
       onChange={(row, [op]) => {
```
If it's not the input table, then the function should immediately return no ?

anisometropie replied:
yeah, made an oversight. it’s better.

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -99,41 +97,64 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
+            rowIndex === 0 ||
+            rowIndex === allWaypoints.length - 1 ||
+            isVia(pathSteps, rowData, WITH_KP),
+        })
```
Nice :)

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -99,41 +97,64 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
+        })
       }
+      cellClassName={({ rowData: rowData_ }) => {
+        const rowData = rowData_ as PathWaypointRow;
```
if isInputTable, add an early return

anisometropie replied:
sure, much better !

clarani commented on front/src/modules/trainschedule/components/DriverTrainScheduleV2/DriverTrainScheduleV2.tsx:
```diff
@@ -28,33 +28,20 @@ const DriverTrainScheduleV2 = ({
   useEffect(() => {
     setBaseOrEco(isEco(simulatedTrain) ? BaseOrEco.eco : BaseOrEco.base);
   }, [simulatedTrain]);
-
```
You can keep this line, it helps to read the code

anisometropie replied:
done

clarani commented on front/src/modules/trainschedule/hooks.ts:
Rename the file useFormattedOperationalPoints

anisometropie replied:
done

clarani commented on front/src/utils/physics.ts:
```diff
@@ -35,4 +35,4 @@ export function mToKmOneDecimal(m: number) {
 }
 
-export const marginRegExValidation = /^\d+(\.\d+)?%$|^\d+(\.\d+)?min\/100km$/;
+export const marginRegExValidation = /^(\d+(\.\d+)?)(%|min\/100km)$/;
```
Move this const closer to where it is used

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,139 @@
+    pathProperties.geometry,
+    100
+  );
+  // const sugOpPointByUic = keyBy(suggestedOperationalPoints, ({ uic, ch }) => `${uic}-${ch}`);
```
```suggestion
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,139 @@
+  pathProperties, // raw from simulation result, calculated values
+  infraId,
+}: TimesStopsOutputProps) => {
+  const simResults = useSimulationResults();
+  usePathfindingV2(); // is it really needed to add position_on_path to pathSteps ? use position_on_path to know where the via is in the suggested point array
```
You should use the trainSchedule given by the parent, as well as the simulation from the parent
```suggestion
```

anisometropie replied:
done for simResults

anisometropie replied:
usePathfindingV2 not needed

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,139 @@
+}: TimesStopsOutputProps) => {
+  const simResults = useSimulationResults();
+  usePathfindingV2(); // is it really needed to add position_on_path to pathSteps ? use position_on_path to know where the via is in the suggested point array
+  const { getPathSteps, getStartTime } = useOsrdConfSelectors();
+  const pathSteps = useSelector(getPathSteps);
+  const startTime = useSelector(getStartTime);
```
Get this from the parent, we try to use the selector only in the views or hooks according to our guidelines

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,139 @@
+        id: selectedTrainId as number,
+      },
+      { skip: !selectedTrainId }
+    );
```
Use the trainSchedule from the parent component

anisometropie replied:
trainSchedule is cached here with rtk, and it’s not directly in the parent SimulationResultsV2

fine for you to leave it ?

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,139 @@
+    100
+  );
+  // const sugOpPointByUic = keyBy(suggestedOperationalPoints, ({ uic, ch }) => `${uic}-${ch}`);
+  const { operationalPoints } = useFormattedOperationalPoints(
+    train,
+    simulatedTrain,
+    pathProperties,
+    infraId
+  );
```
This hook is used in 2 components which are children of SimulationResultsV2. I think it would be better to move it in SimulationResultsV2 and to pass to the children via props

anisometropie replied:
Yes, I don’t think there is a simple way to avoid having  2 calls from the two components, moving it to SimulationResultsV2.tsx

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,139 @@
+    return '';
+  });
+
+  const enrichedoperationalPoints = suggestedOperationalPoints.map((sugOpPoint, sugOpIndex) => {
```
```suggestion
  const enrichedOperationalPoints = suggestedOperationalPoints.map((sugOpPoint, sugOpIndex) => {
```

clarani replied:
Maybe you can create a new hook useOutputTableData which contains all this logic ?

anisometropie replied:
done

clarani commented on front/src/applications/operationalStudies/views/v2/ManageTrainScheduleV2.tsx:
```diff
@@ -107,6 +108,7 @@ const ManageTrainScheduleV2 = () => {
     content: (
       <div className="osrd-config-item-container-map" data-testid="map">
         <div className="floating-itinerary">
+          {/* =========== operational points transformed here */}
```
```suggestion
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -99,41 +95,64 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
+        if (!isInputTable) {
+          return;
+        }
         const rowData = row[`${op.fromRowIndex}`];
```
```suggestion
        const rowData = row[op.fromRowIndex];
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -99,41 +95,64 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
+          return '';
+        }
+        const rowData = rowData_ as PathWaypointRow;
+        const arrivalScheduleNotRespected = rowData.arrival
+          ? rowData.calculatedArrival !== rowData.arrival
+          : false;
```
```suggestion
        const arrivalScheduleNotRespected = !rowData.arrival || rowData.calculatedArrival !== rowData.arrival;
```

anisometropie replied:
not the same truth table,
I need to have false for `rowData.arrival` `undefined`

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -99,41 +95,64 @@ const TimesStops = ({ pathProperties, pathSteps = [], startTime }: TimesStopsPro
+        const arrivalScheduleNotRespected = rowData.arrival
+          ? rowData.calculatedArrival !== rowData.arrival
+          : false;
+        const negativeDiffMargins = Number(rowData?.diffMargins?.split(NO_BREAK_SPACE)[0]) < 0;
```
```suggestion
        const negativeDiffMargins = Number(rowData.diffMargins?.split(NO_BREAK_SPACE)[0]) < 0;
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,47 @@
+  pathProperties: PathPropertiesFormatted;
+  operationalPoints: OperationalPointWithTimeAndSpeed[];
+  startTime: string;
+  pathSteps: (PathStep | null)[];
```
You can compact the pathSteps in the parent component (null value are only if the path is not fully defined, which is not the case for simulation results). You can also remove the compact line 40.
```suggestion
  pathSteps: PathStep[];
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,79 @@
+
+let marginUnit = '%';
+
+const useComputedMargins = (
```
Actually it's simply a helper, so maybe we can transform it into a helper function (remove 'use' in front of its name), move it to a folder helpers and add some tests

anisometropie replied:
how is this a helper if useOutputTableData is a hook ?
it contains even more business rules and logic

SharglutDev replied:
I think what @clarani means is that a custom hook (useXXX) should be used if it uses some hooks itself (useState, useEffect, a rtk query call...) but that's not the case here. It's just a large function so we don't need to call it "use"ComputedMargins.

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,79 @@
+let marginUnit = '%';
+
+const useComputedMargins = (
+  trainSimulation: SimulationResponse | undefined,
```
```suggestion
  trainSimulation: SimulationResponseSuccess,
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,79 @@
+  selectedTrainSchedule: TrainScheduleResult,
+  pathId: string
+) => {
+  if (trainSimulation?.status === 'success') {
```
The simulation is always a success (see in the parent hook), so you can remove this test
```suggestion
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,79 @@
+  pathId: string
+) => {
+  if (trainSimulation?.status === 'success') {
+    const theoreticalMarginIndex = selectedTrainSchedule?.margins?.boundaries?.findIndex(
```
You can test above if margins are defined, otherwise return the default values and also remove the ?
```suggestion
    const theoreticalMarginIndex = selectedTrainSchedule.margins.boundaries.findIndex(
```

anisometropie replied:
we have two reasons for failures, margins can be undefined, but we also can find no index. 
we can put that in a single if (!theoreticalMarginIndex) to handle both

anisometropie replied:
```
  const theoreticalMarginIndex = selectedTrainSchedule?.margins?.boundaries?.findIndex(
    (id) => id === pathId
  );
  if (!theoreticalMarginIndex) {
    return ['', '', ''];
  }
```

clarani replied:
Yes I agree :) You can also remove the ? after selectedTrainSchedule

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,111 @@
+      { skip: !selectedTrainId }
+    );
+
+  const trainScheduleByAt = keyBy(selectedTrainSchedule.schedule, 'at');
```
I thought at first that schedulesByAt was containing trainSchedules and not only schedules
```suggestion
  const schedulesByAt = keyBy(selectedTrainSchedule.schedule, 'at');
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,111 @@
+      formatSuggestedOperationalPoints(
+        pathProperties.operationalPoints,
+        pathProperties.geometry,
+        100
```
Why 100 and not pathProperties.length ?
```suggestion
        pathProperties.length
```

anisometropie replied:
fixed

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,111 @@
+    [pathProperties.operationalPoints, pathProperties.geometry]
+  );
+
+  const pathPointsWithOpPointIndices = selectedTrainSchedule.path.map((pathPoint) => {
```
Use pathStep instead (it is the term we use every where else). Replace here, and everywhere else in this file

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,111 @@
+  );
+
+  const pathPointsWithOpPointIndices = selectedTrainSchedule.path.map((pathPoint) => {
+    const correspondingOpPointIndex = suggestedOperationalPoints.findIndex((sugOpPoint) => {
+      if ('uic' in pathPoint) {
+        return sugOpPoint.uic === pathPoint.uic && sugOpPoint.ch === pathPoint.secondary_code;
+      }
+      return false;
+    });
```
```suggestion
    const correspondingOpPointIndex = suggestedOperationalPoints.findIndex(
      (sugOpPoint) =>
        'uic' in pathStep &&
        sugOpPoint.uic === pathStep.uic &&
        sugOpPoint.ch === pathStep.secondary_code
    );
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,111 @@
+      return pathPoint.operational_point;
+    }
+    return '';
+  });
```
You can filter the pathSteps without uic above (in pathPointsWithOpPointIndices) and then use this instead:
```suggestion
  const trainSchedulePathByUic = keyBy(pathStepsWithOpPointIndices, (pathStep) =>
    'uic' in pathStep ? `${pathStep.uic}-${pathStep.secondary_code}` : ''
  );
   );
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,111 @@
+  });
+
+  return suggestedOperationalPoints.map((sugOpPoint, sugOpIndex) => {
+    const opPoint = operationalPoints[sugOpIndex];
```
How can you know that opPoint corresponds to sugOpPoint ?

anisometropie replied:
i made a lookup table operationPointsByNameCh

clarani commented on front/src/modules/timesStops/utils.ts:
```diff
@@ -49,3 +53,36 @@ export const formatSuggestedViasToRowVias = (
+
+export function findNextScheduledPoint(
+  operationalPoints: OperationalPointWithTimeAndSpeed[],
+  tsPathPointsWithOpPointIndices: PathPointOpPointCorrespondance[],
```
Why is there a ts prefix ? Also use pathStep instead of pathPoint. You can also rename the type PathPointOpPointCorrespondance
```suggestion
  pathStepsWithOpIndex: PathPointOpPointCorrespondance[],
```

anisometropie replied:
done

clarani commented on front/src/applications/operationalStudies/views/v2/ManageTrainScheduleV2.tsx:
```diff
@@ -134,10 +134,11 @@ const ManageTrainScheduleV2 = ({ trainIdToEdit }: ManageTrainScheduleV2Props) =>
+        allWaypoints={pathProperties?.allWaypoints}
         startTime={startTime}
+        pathSteps={compact(pathSteps)}
+        // tableType={TableType.Input}
```
```suggestion
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -1,63 +1,43 @@
   startTime?: string;
+  tableType: TableType;
+  cellClassName?: DataSheetGridProps['cellClassName'];
+  stickyRightColumn?: DataSheetGridProps['stickyRightColumn'];
```
You can even move the onChange function in a props:
```suggestion
  stickyRightColumn?: DataSheetGridProps['stickyRightColumn'];
  onChange?: (
    row: PathWaypointRow[],
    op: Operation,
    setRows: (rows: PathWaypointRow[]) => void
  ) => void;
```

anisometropie replied:
leaving it as discussed

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+      correspondingOpPointIndex,
+    };
+  });
+  const trainSchedulePathByUic = keyBy(pathStepsWithOpPointIndices, (pathStep) => {
```
```suggestion
  const pathStepsByUic = keyBy(pathStepsWithOpPointIndices, (pathStep) => {
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+      correspondingOpPointIndex,
+    };
+  });
+  const trainSchedulePathByUic = keyBy(pathStepsWithOpPointIndices, (pathStep) => {
+    if ('uic' in pathStep) {
+      return `${pathStep.uic}-${pathStep.secondary_code}`;
+    }
+    if ('operational_point' in pathStep) {
+      return pathStep.operational_point;
+    }
+    return '';
+  });
```
```suggestion
  const trainSchedulePathByUic = keyBy(pathStepsWithOpPointIndices, (pathStep) =>
    'uic' in pathStep ? `${pathStep.uic}-${pathStep.secondary_code}` : ''
  );
```

anisometropie replied:
done, with the filter the `'uic' in pathStep` check is gone too

clarani commented on front/src/modules/timesStops/utils.ts:
```diff
@@ -43,3 +47,36 @@ export const formatSuggestedViasToRowVias = (
+  const pathIndex = tsPathPointsWithOpPointIndices.findIndex(
+    (pathPoint) => pathPoint.correspondingOpPointIndex === sugOpIndex
+  );
+  const nextPathPoint = tsPathPointsWithOpPointIndices[pathIndex + 1];
```
You should test here that pathIndex is not -1. Also, it should be pathStep index

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/utils.ts:
```diff
@@ -43,3 +47,36 @@ export const formatSuggestedViasToRowVias = (
   });
 };
+
+export function findNextScheduledPoint(
```
```suggestion
export function findNextScheduledPathStep(
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+      const arrival = schedule?.arrival ? ISO8601Duration2sec(schedule.arrival!) : null;
+      const stopFor = schedule?.stop_for ? ISO8601Duration2sec(schedule.stop_for!) : '';
+      const departure = arrival && stopFor ? secToHoursString(arrival + stopFor) : null;
+      const [theoreticalMargin, realMargin, diffMargins] = useComputedMargins(
```
Use a desconsctruction instead of relying on the position in the order.
```suggestion
      const { theoreticalMargin, realMargin, diffMargins } = useComputedMargins(
```
You can also store the value directly in marginsData and use it like this:
```
return {
    ...,
    ...marginsData
}
```

anisometropie replied:
done.
I think it’s nice to have a visual of the entire object being returned.

clarani replied:
As you prefer

anisometropie replied:
i did the `...marginsData` and `...scheduleData` eventually

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+      pathStepsWithOpPointIndices,
+      sugOpIndex
+    );
+    const pathKey = `${sugOpPoint?.uic}-${sugOpPoint?.ch}`;
```
```suggestion
    const pathKey = `${sugOpPoint.uic}-${sugOpPoint.ch}`;
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+      sugOpIndex
+    );
+    const pathKey = `${sugOpPoint?.uic}-${sugOpPoint?.ch}`;
+    const pathStep = trainSchedulePathByUic[pathKey];
```
Move this line after the if l72

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+    const pathStep = trainSchedulePathByUic[pathKey];
+
+    if (pathKey in trainSchedulePathByUic) {
+      const pathId = 'id' in pathStep ? pathStep.id : '';
```
All the variant have an id. Since pathStep is only used here, I think you don't need to instantiate it
```suggestion
      const pathId = pathStepsByUic[pathKey].id
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+      pathStepsWithOpPointIndices,
+      sugOpIndex
+    );
+    const pathKey = `${sugOpPoint?.uic}-${sugOpPoint?.ch}`;
```
What do you think of pathStepKey ?
```suggestion
    const pathStepKey = `${sugOpPoint?.uic}-${sugOpPoint?.ch}`;
```

anisometropie replied:
sure

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+
+    if (pathKey in trainSchedulePathByUic) {
+      const pathId = 'id' in pathStep ? pathStep.id : '';
+      const schedule = scheduleByAt[pathId];
+      const arrival = schedule?.arrival ? ISO8601Duration2sec(schedule.arrival!) : null;
+      const stopFor = schedule?.stop_for ? ISO8601Duration2sec(schedule.stop_for!) : '';
+      const departure = arrival && stopFor ? secToHoursString(arrival + stopFor) : null;
```
All these props is a bit difficult to read. I think it would be easier if you extract all the code relative to schedule in a utils function:
```suggestion
      const schedule = scheduleByAt[pathId];
      const scheduleData = formatScheduleData(schedule, startTime)
```
and then in the return: 
```
return {
        ...sugOpPoint,
        // calculated arrival/departure
        calculatedArrival: secToHoursString(opPoint?.time, true),
        calculatedDeparture:
          opPoint?.duration > 0 ? secToHoursString(opPoint?.time + opPoint?.duration, true) : '',
        
       // schedule data
        ...scheduleData,
       
        // margins data
        ...marginsData
      } as SuggestedOP;
```

anisometropie replied:
nice, done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+        departure,
+        stopFor: String(stopFor), // I have some numbers in the input table state
+        onStopSignal: schedule?.on_stop_signal || '',
+        calculatedArrival: secToHoursString(opPoint?.time, true),
+        calculatedDeparture:
+          opPoint?.duration > 0 ? secToHoursString(opPoint?.time + opPoint?.duration, true) : '',
```
Add an early return if opPoint is undefined, and remove the ?
```suggestion
        calculatedArrival: secToHoursString(opPoint.time, true),
        calculatedDeparture:
          opPoint.duration ? secToHoursString(opPoint.time + opPoint.duration, true) : '',
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+        departure,
+        stopFor: String(stopFor), // I have some numbers in the input table state
+        onStopSignal: schedule?.on_stop_signal || '',
+        calculatedArrival: secToHoursString(opPoint?.time, true),
```
Do you think you can use a const to replace the magic true here ?

anisometropie replied:
it’s not magic as if you had random numbers or strings in your code. 
I don’t really want to have a const everytime a function expects a boolean. the function’s params are already named.

clarani commented on front/src/modules/timesStops/styles/_timesStopsDatasheet.scss:
```diff
@@ -1,19 +1,32 @@
-  color: red !important;
-}
+  .warning-schedule {
+    // warning-30
+    // @extend .text-warning-30;
```
Remove the comments
```suggestion
```

anisometropie replied:
I’m leaving warning-30, we need to know what color it is from sketch colors when the variables become available

clarani commented on front/src/modules/timesStops/styles/_timesStopsDatasheet.scss:
```diff
@@ -1,19 +1,32 @@
-  button {
-    margin: auto;
+  .warning-margin {
+    // warning-30 50% alpha
```
```suggestion
```

anisometropie replied:
we need to know what color it is from sketch colors when we have the variables

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,102 @@
+      } as SuggestedOP;
+    }
+
+    return { ...sugOpPoint, calculatedArrival: secToHoursString(opPoint?.time, true) };
```
Remove the ? if you have added an early return
```suggestion
    return { ...sugOpPoint, calculatedArrival: secToHoursString(opPoint.time, true) };
```

anisometropie replied:
yes

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+    (id) => id === pathId
+  );
+  if (!theoreticalMarginIndex) {
+    return ['', '', ''];
```
I think it would be better to return an object like this, what do you think ?
```suggestion
    return {};
```

anisometropie replied:
yes, with the empty strings `{ theoreticalMarginWithUnit: '', calculatedMargin: '', diffMarginsWithUnit: '' }`

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+  }
+  const theoreticalMargin = selectedTrainSchedule.margins!.values[theoreticalMarginIndex + 1];
+  const theoMarginRegexMatches = theoreticalMargin?.match(marginRegExValidation);
+  marginUnit = theoMarginRegexMatches?.[3] || marginUnit;
```
Remove the let l9
```suggestion
  const marginUnit = theoMarginRegexMatches?.[3] || '%';
```

anisometropie replied:
`marginUnit` is kept between the function calls to have use the same unit as the previous column or row

clarani replied:
Ok my bad, I missed that, thanks !

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+  const opPointBaseTime = interpolateValue(trainSimulation.base, opPoint?.position * 1000, 'times');
+  const nextOpPointBaseTime = interpolateValue(
+    trainSimulation.base,
+    nextOpPoint?.position * 1000,
```
Remove all the ? which are not necessary and use the utils function mToMm instead of `* 1000`
```suggestion
    nextOpPoint.position * 1000,
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+    'times'
+  );
+  const baseDuration = nextOpPointBaseTime - opPointBaseTime;
+  const opPointprovTime = interpolateValue(
```
Use the full word here, since it is not a word we see everyday
```suggestion
  const opPointprovisionalTime = interpolateValue(
```

anisometropie replied:
done for both opPointprovisionalTime nextOpPointProvisionalTime

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+  const theoreticalMarginIndex = selectedTrainSchedule?.margins?.boundaries?.findIndex(
+    (id) => id === pathId
+  );
+  if (!theoreticalMarginIndex) {
```
Early return if:
- !theoreticalMarginIndex
- theoreticalMarginIndex is -1 or is the last one
- selectedTrainSchedule has no margins (maybe this test can be done before the call to the function)

anisometropie replied:
done.
if we have no `selectedTrainSchedule.margins` we know theoreticalMarginIndex will be undefined as well

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+    return ['', '', ''];
+  }
+  const theoreticalMargin = selectedTrainSchedule.margins!.values[theoreticalMarginIndex + 1];
+  const theoMarginRegexMatches = theoreticalMargin?.match(marginRegExValidation);
```
Early return if theoMarginRegexMatches is false or if theroreticalMargin is undefined

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+    }
+  }
+
+  const theoreticalMarginWithUnit = formatDigitsAndUnit(theoreticalMargin);
```
Can you re-use theoreticalMargin directly ?

anisometropie replied:
We need a non-breaking space to abide by typography rules

100 km
100 %

clarani replied:
Understood

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+  if (marginUnit === '%') {
+    realMargin = 100 * (realLostTime / baseDuration) || 0;
+    if (theoreticalMargin) {
+      const theoreticalMarginValue = Number(theoMarginRegexMatches?.[1]);
```
Move this line under `const marginUnit = theoMarginRegexMatches[3];` l26 and remove it here and l59
```suggestion
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
- remove all the ? not necessary
- remove all the magic numbers and use utils functions to convert in the correct unit
- add blank lines and comments if needed to make the code clearer and easier to read

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useComputedMargins.tsx:
```diff
@@ -0,0 +1,73 @@
+  let diffMargins = 0;
+  if (marginUnit === '%') {
+    realMargin = 100 * (realLostTime / baseDuration) || 0;
+    if (theoreticalMargin) {
```
Remove this line as you know that theoreticalMargin is defined (same l60). 
Thus diffMargins is always defined too and you can remove the test l69

anisometropie replied:
done.

we need this check if it’s `0`
`diffMargins ? formatDigitsAndUnit(String(diffMargins), 's') : ''`

clarani commented on front/src/modules/timesStops/types.ts:
```diff
@@ -1,5 +1,12 @@
+  Output = 'Output',
+}
+
+export type PathPointOpPointCorrespondance = { correspondingOpPointIndex: number };
```
Use PathStep instead of pathPoint. Also maybe opPoint is a bit redundant, op (for operation point) is not enough ? (only a suggestion)
```suggestion
export type PathStepOpPointCorrespondance = { correspondingOpPointIndex: number };
```

anisometropie replied:
done. I guess `opPoint` is clearer

clarani commented on front/src/modules/timesStops/utils.ts:
```diff
@@ -43,3 +47,36 @@ export const formatSuggestedViasToRowVias = (
+  return operationalPoints[nextPathPoint?.correspondingOpPointIndex];
+}
+
+const getDigits = (u: string | undefined) => (u === 's' || u === '%' ? 0 : 1);
```
Maybe getDigitsNumber is clearer. Also, you can rename u `unit`
```suggestion
const getDigitsNumber = (unit: string | undefined) => (unit === 's' || unit === '%' ? 0 : 1);
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/utils.ts:
```diff
@@ -43,3 +47,36 @@ export const formatSuggestedViasToRowVias = (
+  if (value === undefined || value === 'none') {
+    return '';
+  }
+  if (marginRegExValidation.test(value)) {
+    const [, v, , u] = value.match(marginRegExValidation) || [];
```
Why not directly match like in useComputedMargins ?
```suggestion
  const valueSplit = value.match(marginRegExValidation);
  if (valueSplit) {
    const [, v, , u] = valueSplit;
```

anisometropie replied:
ah yes, done

clarani commented on front/src/modules/timesStops/utils.ts:
```diff
@@ -43,3 +47,36 @@ export const formatSuggestedViasToRowVias = (
+    return '';
+  }
+  if (marginRegExValidation.test(value)) {
+    const [, v, , u] = value.match(marginRegExValidation) || [];
```
Give a real name for v and u

clarani commented on front/src/modules/timesStops/utils.ts:
```diff
@@ -43,3 +47,36 @@ export const formatSuggestedViasToRowVias = (
+
+const getDigits = (u: string | undefined) => (u === 's' || u === '%' ? 0 : 1);
+
+export function formatDigitsAndUnit(value: string | undefined, unit?: string) {
```
Actually, since this function is only used three times, I think you can remove the call with theoreticalMargin in useComputedMargins and simplify the function by removing (change also the signature of the function)  
```suggestion
export function formatDigitsAndUnit(value: number | undefined, unit: string) {
```

anisometropie replied:
we need a space between a value and unit for `theoreticalMargin`

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,66 @@
+  selectedTrainSchedule: TrainScheduleResult;
+  startTime: string;
+  pathSteps: PathStep[];
+  pathLength?: number;
```
I think pathLength should be not nullable

anisometropie replied:
useSimulationResults should handle this?

clarani reviewed (commented):
Great job :)

clarani commented on front/src/modules/timesStops/styles/_timesStopsDatasheet.scss:
```diff
@@ -1,19 +1,31 @@
-  color: red !important;
-}
+  .warning-schedule {
+    // warning-30
```
What does this comment mean ? same for the one l27
```suggestion
```

anisometropie replied:
names from sketch, we to replace the hex values when we have the variables
![Screenshot_20240718_130400](https://github.com/user-attachments/assets/e5a72cfc-8b86-40fd-b554-68bcde1a2229)

clarani commented on front/src/modules/timesStops/types.ts:
```diff
@@ -1,5 +1,19 @@
+export type PathStepOpPointCorrespondance = { correspondingOpPointIndex: number };
+
+type TrainSchedulePath = TrainScheduleBase['path'][number];
+export type TrainScheduleBasePathWithUic = Extract<TrainSchedulePath, { uic: number }>;
```
Super nice !

clarani commented on front/src/modules/timesStops/types.ts:
```diff
@@ -1,5 +1,19 @@
+
+export type PathStepOpPointCorrespondance = { correspondingOpPointIndex: number };
+
+type TrainSchedulePath = TrainScheduleBase['path'][number];
```
```suggestion
type TrainSchedulePathStep = ArrayElement<TrainScheduleBase['path']>;
```

anisometropie replied:
done

clarani commented on front/src/utils/__tests__/physics.spec.ts:
```diff
@@ -9,3 +9,11 @@ describe('jouleToKwh', () => {
+describe('minutePer100km', () => {
+  it('should return margin in min/100km', () => {
+    const seconds = 60;
+    const metres = 1000;
```
```suggestion
    const meters = 1000;
```

clarani commented on front/src/utils/date.ts:
```diff
@@ -10,9 +10,16 @@ import { ISO8601Duration2sec } from './timeManipulation';
 dayjs.extend(utc);
 dayjs.extend(timezone);
 
+export function extractHHMMSS(dateTimeString?: string) {
```
Maybe you can add a comment to specify that the dateTimeString is in ISO format ?

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/formatScheduleData.ts:
```diff
@@ -0,0 +1,16 @@
+import type { ScheduleEntry } from '../types';
+
+function formatScheduleData(schedule: ScheduleEntry, startTime: string) {
+  const arrival = schedule?.arrival ? ISO8601Duration2sec(schedule.arrival!) : null;
```
Add an early return and remove the ? after schedule
```suggestion
    if (!schedule) return ''
    const arrival = schedule.arrival ? ISO8601Duration2sec(schedule.arrival!) : null;
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,108 @@
+  opPoint: OperationalPointWithTimeAndSpeed,
+  nextOpPoint: OperationalPointWithTimeAndSpeed,
+  selectedTrainSchedule: TrainScheduleResult,
+  pathId: string
```
```suggestion
  pathStepId: string
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,108 @@
+  theoreticalMarginWithUnit: '',
+  calculatedMargin: '',
+  diffMarginsWithUnit: '',
+};
```
It's a detail, but don't hesitate to add blank lines to make the code easier to read
```suggestion
};

```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,108 @@
+    mToMm(nextOpPoint.position),
+    'times'
+  );
+  const baseDuration = milliUnitToBaseUnit(nextOpPointBaseTime - opPointBaseTime);
```
milliUnitToBaseUnit is a name quite abstract, I think msToS is clearer
```suggestion
  const baseDuration = msToS(nextOpPointBaseTime - opPointBaseTime);
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,108 @@
+
+  // how much longer it took (s) with the margin than without
+  const realLostTime = provisionalDuration - baseDuration;
+  const distanceToNextOpPoint = nextOpPoint.position - opPoint.position; // metre
```
```suggestion
  const distanceToNextOpPoint = nextOpPoint.position - opPoint.position; // meter
```

flomonster reviewed (commented):
I tested it adding a single margin. The "Theoritical margin" is fully empty.

clarani commented on front/src/modules/timesStops/helpers/utils.ts:
```diff
@@ -0,0 +1,93 @@
+import { TableType } from '../types';
+import type { PathStepOpPointCorrespondance, PathWaypointRow } from '../types';
+
+// eslint-disable-next-line import/prefer-default-export
```
```suggestion
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/utils.ts:
```diff
@@ -0,0 +1,93 @@
+  });
+};
+
+export function findNextScheduledPathStep(
```
```suggestion
export function findNextPathStep(
```

anisometropie replied:
actually it’s returning an element of `operationalPoints` , the next that corresponds to the next `pathStep`.
perhaps with can keep findNextScheduledOpPoint ?

clarani replied:
Yes I agree

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/utils.ts:
```diff
@@ -0,0 +1,93 @@
+  pathStepsWithOpPointIndex: PathStepOpPointCorrespondance[],
+  sugOpIndex: number
+) {
+  const pathIndex = pathStepsWithOpPointIndex.findIndex(
```
```suggestion
  const pathStepIndex = pathStepsWithOpPointIndex.findIndex(
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/helpers/utils.ts:
```diff
@@ -0,0 +1,93 @@
+    (pathStep) => pathStep.correspondingOpPointIndex === sugOpIndex
+  );
+  if (pathIndex >= 0) {
+    const nextPathPoint = pathStepsWithOpPointIndex[pathIndex + 1];
```
```suggestion
    const nextPathStep = pathStepsWithOpPointIndex[pathIndex + 1];
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
```diff
@@ -0,0 +1,117 @@
+        const pathStepKey = `${sugOpPoint.uic}-${sugOpPoint.ch}`;
+
+        if (pathStepKey in pathStepsByUic && nextOpPoint) {
+          const pathId = pathStepsByUic[pathStepKey].id || '';
```
```suggestion
          const pathStepId = pathStepsByUic[pathStepKey].id || '';
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/types.ts:
```diff
@@ -1,5 +1,19 @@
+
+export type PathStepOpPointCorrespondance = { correspondingOpPointIndex: number };
+
+type TrainSchedulePath = ArrayElement<TrainScheduleBase['path']>;
```
```suggestion
type TrainSchedulePathStep = ArrayElement<TrainScheduleBase['path']>;
```

anisometropie replied:
done

anisometropie committed:
front: add simulation result stops output table

This commit implements a new component to display the simulation results in a summary table.
Stops, margins, arrival time request before the simulation are displayed, and compared
with the actual stop, margins, times we got during the simulation.
The table will highlight the train stops where too much discrepancy is observed.

clarani commented on front/src/modules/timesStops/TimesStops.tsx:
```diff
@@ -1,63 +1,43 @@
   const { t } = useTranslation('timesStops');
-
-  if (!pathProperties) {
+  if (!allWaypoints) {
```
You should do the return after the useEffect, otherwise I encounter this error when clicking too quickly on the times and stops tab
![image](https://github.com/user-attachments/assets/c1ab373d-b6fe-466b-9f43-4bdb296b5d61)

clarani replied:
This bug is already on dev and should be fixed in another PR: https://github.com/OpenRailAssociation/osrd/issues/8114 :+1:

clarani reviewed (approved):
LGTM, tested :white_check_mark:

flomonster commented on front/public/locales/en/timesStops.json:
```diff
@@ -1,10 +1,15 @@
 {
+  "arrivalTime": "Asked arrival Time",
+  "calculatedArrivalTime": "Calculated arrival time",
+  "calculatedDepartureTime": "Calculated departure time",
+  "departureTime": "Asked departure Time",
+  "diffMargins": "Diff. margins",
```
```suggestion
  "arrivalTime": "Requested arrival time",
  "calculatedArrivalTime": "Calculated arrival time",
  "calculatedDepartureTime": "Calculated departure time",
  "departureTime": "Requested departure time",
  "diffMargins": "Margins diff.",
```

anisometropie replied:
done

flomonster reviewed (changes requested):
I found some bugs / user-interface issues:


## User interface

![image](https://github.com/user-attachments/assets/dd7d4567-3978-4940-a476-99df5594c624)

1. Text is cut off by the height of the box
2. Alignment is not consistent

## Bugs

The real margin can be less than the theoretical margin and the difference is still `0`.

![image](https://github.com/user-attachments/assets/e9bcc3ba-d65a-4875-82a1-87d7dde27d7b)

SharglutDev commented on front/src/modules/timesStops/consts.ts:
```diff
@@ -0,0 +1,7 @@
+/* eslint-disable import/prefer-default-export */
```
I don't think you need this anymore

anisometropie replied:
done

SharglutDev commented on front/src/modules/timesStops/TimesStopsInput.tsx:
```diff
@@ -0,0 +1,89 @@
+  if (isRowVia) {
+    return (
+      <button type="button" onClick={removeVia}>
+        ❌
```
@thibautsailly would prefer that we don't use a red cross here but a primary cross from ui-icons

anisometropie replied:
done

anisometropie replied:
reverted

SharglutDev commented on front/src/modules/timesStops/hooks/useOutputTableData.tsx:
This should be a `.ts` file

anisometropie replied:
done

SharglutDev commented on front/src/modules/timesStops/TimeStopsColumns.tsx:
Should this file be renamed `useTimeStopsColumns.ts` in a hooks folder ?

anisometropie replied:
yes, done

eckter commented on front/src/applications/operationalStudies/views/v2/SimulationResultsV2.tsx:
```diff
@@ -187,14 +203,31 @@ const SimulationResultsV2 = ({
         </div>
       )}
 
+      {selectedTrainSchedule &&
```
This section doesn't have a `{/* */}` comment before like the other ones on the page

anisometropie replied:
done

eckter commented on front/src/utils/timeManipulation.ts:
```diff
@@ -87,6 +87,10 @@ export function ISO8601Duration2sec(isoDuration: string) {
 
-export function getStopTime(sec: number) {
-  return new Date(sec * 1000).toISOString().substr(11, 5);
+export function secToHoursString(sec: number | null, withSeconds = false) {
```
Could use a comment, its purpose is not entirely clear. And I believe it's missing a type annotation for the return? (though that seems common in the code base, maybe it's normal)

anisometropie replied:
done.
we don’t usually specify the return type annotation as it’s inferred. Here we have an alias, so we might as well use it for the return type.

eckter replied:
Thanks, it's easier to understand :+1: 

But now that I get the intent, I feel like using `d3.timeFormat('%H:%M')` in the function would make it easier to follow (it's used in other places in the file).

But feel free to ignore this if you're running out of time, I realize I'm not making things go faster here

anisometropie replied:
done

eckter commented on front/src/utils/timeManipulation.ts:
```diff
@@ -87,6 +87,10 @@ export function ISO8601Duration2sec(isoDuration: string) {
+    return '';
+  }
+  const length = withSeconds ? 8 : 5;
+  return new Date(sec * 1000).toISOString().substr(11, length);
```
substr appears to be deprecated

anisometropie replied:
changed to `substring`

eckter commented on front/src/modules/timesStops/consts.ts:
```diff
@@ -0,0 +1,6 @@
+export const marginRegExValidation = /^(\d+(\.\d+)?)(%|min\/100km)$/;
+
+export const MarginUnit = {
+  percent: '%',
+  second: 's',
+} as const;
```
Is there a reason not to have `min/100km`? That feels wrong, but I don't really get the full context

anisometropie replied:
It wasn’t used, but ok, I added it.

eckter commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,94 @@
+    if (
+      theoriticalMarginBoundaryIndex === undefined ||
+      theoriticalMarginBoundaryIndex < 0 ||
+      theoriticalMarginBoundaryIndex > selectedTrainSchedule.margins!.values.length - 2
+    ) {
+      return undefined;
+    }
+    theoreticalMarginIndex = theoriticalMarginBoundaryIndex + 1;
+  }
+  return selectedTrainSchedule.margins!.values[theoreticalMarginIndex];
```
Assuming the check `... > length - 2` is there to avoid the out of bounds when getting `[index + 1`], the condition is too restrictive by 1.

You can ignore this comment if we do actually want to return undefined on the last value.

anisometropie replied:
`boundaries` length is N, `values` length is N + 1.

`boundaryIndex + 1` is out of bounds in `values` if `boundaryIndex + 1 > values.length - 1`

so the condition `boundariesIndex> values.length - 2` is looking fine to me. 

am I missing something ?

eckter replied:
I got confused on the boundaries vs values, all good

eckter commented on front/src/modules/timesStops/helpers/formatScheduleData.ts:
```diff
@@ -0,0 +1,23 @@
+  return {
+    arrival: arrivalSeconds ? secToHoursString(startTimeSeconds + arrivalSeconds, true) : '',
+    departure,
+    stopFor: String(stopForSeconds), // I have some numbers in the input table state
```
A comment written in first-person feels wrong

The value doesn't come from the input table state, right? It's defined at line 11 (parsed from a string)

anisometropie replied:
removed it as it’s no longer relevant

eckter reviewed (commented):
I haven't checked everything yet. I'm discovering the code base in the front-end (and react in general) so the comments may not be 100% relevant

Note : the output table only contains operational points and not waypoints added by hand. Is that normal?

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,94 @@
+const emptyMargins = {
+  theoreticalMargin: '',
+  theoreticalMarginSeconds: '',
+  calculatedMarginSeconds: '',
+  diffMarginsSeconds: '',
```
```suggestion
  calculatedMargin: '',
  diffMargins: '',
```

anisometropie replied:
well it’s not used anymore

anisometropie replied:
it,s not used anymore

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,94 @@
+  pathStepId: string
+) {
+  const theoreticalMargin = getTheoreticalMargin(selectedTrainSchedule, pathStepId);
+  if (!theoreticalMargin) {
+    return emptyMargins;
+  }
+  if (!theoreticalMargin || theoreticalMargin === '0%') {
+    return emptyMargins;
+  }
```
Since we want to display the margin in seconds even if no margin theoritical is set (but we should check whether the pathStep has an arrivalTime or whether it is a stop)
```suggestion
if ((!theoreticalMargin|| theoreticalMargin === '0%') && !pathStepIsAStop && !pathStepHasArrivalTime) {
    return emptyMargins;
  }
```

anisometropie replied:
I think we never need to return early, not anymore.

clarani replied:
You're right, since we don't have the via added by clicks on the map :+1:

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,94 @@
+  let theoreticalMarginIndex;
+  if (pathStepIndex === 0) {
+    theoreticalMarginIndex = 0;
+  } else {
```
You can replace your loop findIndex by 2 tests, which will be more efficient and optimized
```suggestion
  if (selectedTrainSchedule.path.length === 0) return undefined;

  if (selectedTrainSchedule.path[0].id === pathStepId) {
    return selectedTrainSchedule.margins?.values[0];
  }
```

anisometropie replied:
great change, thanks

clarani commented on front/src/modules/timesStops/helpers/computeMargins.tsx:
```diff
@@ -0,0 +1,88 @@
+  // base = no margin
+  // provisional = margins
+  // final = margin + requested arrival times
+  const baseDuration = computeDuration(trainSimulation, opPoint, nextOpPoint, 'base');
```
And what if you change the signature of the computeDuration function like this (use ReportTrainV2 directly) ?
```
function computeDuration(
  trainSimulation: ReportTrainV2,
  opPoint: OperationalPointWithTimeAndSpeed,
  nextOpPoint: OperationalPointWithTimeAndSpeed
) {}
```
and then:
```suggestion
  const baseDuration = computeDuration(trainSimulation.base, opPoint, nextOpPoint);
```

anisometropie replied:
good stuff, done

clarani commented on front/src/modules/timesStops/hooks/useTimeStopsColumns.tsx:
You can even rename it .ts

anisometropie replied:
done

clarani commented on front/src/utils/timeManipulation.ts:
```diff
@@ -87,6 +87,15 @@ export function ISO8601Duration2sec(isoDuration: string) {
+ * converts a value in seconds to a timeString "HH:MM"
+ *
+ * using the param withSeconds returns the longer format "HH:MM:SS"
+ */
```
Thanks for the comment :pray:

flomonster reviewed (approved):
It's working but I found small things to adapt:
- we should have a threshold where we snap the calculated arrival time to the requested arrival time (1s of diff is okay)
- Columns title is cropped by the box size

> [!NOTE]
> Tested but not reviewed

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,66 @@
+    pathLength
+  );
+  return (
+    <TimesStops
+      allWaypoints={enrichedOperationalPoints}
+      pathSteps={pathSteps}
+      startTime={startTime}
+      tableType={TableType.Output}
+      cellClassName={({ rowData: rowData_ }) => {
+        const rowData = rowData_ as PathWaypointRow;
+        const arrivalScheduleNotRespected = rowData.arrival
+          ? rowData.calculatedArrival !== rowData.arrival
+          : false;
+        const negativeDiffMargins = Number(rowData.diffMargins?.split(NO_BREAK_SPACE)[0]) < 0;
+        return cx({
+          'warning-schedule': arrivalScheduleNotRespected,
+          'warning-margin': negativeDiffMargins,
+        });
+      }}
+    />
```
Small improvement: currently, if I select a new train, I can still see for a moment the results of the previous train and there is nothing to notify me that the data has finished to load. Maybe we can add a loader like this (you can reuse formattedOpPointsLoading in SimulationResultsV2)
```suggestion
    // Prevent the screen from resizing during loading
    <div style={{ height: '600px' }}>
      {(dataIsLoading && <Loader />) || (
        <TimesStops
          allWaypoints={enrichedOperationalPoints}
          pathSteps={pathSteps}
          startTime={startTime}
          tableType={TableType.Output}
          cellClassName={({ rowData: rowData_ }) => {
            const rowData = rowData_ as PathWaypointRow;
            const arrivalScheduleNotRespected = rowData.arrival
              ? rowData.calculatedArrival !== rowData.arrival
              : false;
            const negativeDiffMargins = Number(rowData.diffMargins?.split(NO_BREAK_SPACE)[0]) < 0;
            return cx({
              'warning-schedule': arrivalScheduleNotRespected,
              'warning-margin': negativeDiffMargins,
            });
          }}
        />
      )}
    </div>
```

anisometropie replied:
done

clarani commented on front/src/applications/operationalStudies/views/v2/SimulationResultsV2.tsx:
```diff
@@ -187,14 +203,32 @@ const SimulationResultsV2 = ({
+              pathProperties={pathProperties}
+              operationalPoints={operationalPoints}
+              selectedTrainSchedule={selectedTrainSchedule}
+              startTime={startTime}
```
StartTime should be the selectedTrainSchedule start_time, and not the one kept in the store (it's the one used to ManageTrainSchedule, to create or edit a train)
```suggestion
              startTime={selectedTrainSchedule.start_time}
```

clarani commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -0,0 +1,66 @@
+    <TimesStops
+      allWaypoints={enrichedOperationalPoints}
+      pathSteps={pathSteps}
+      startTime={startTime}
```
Actually, you can use directly the start time of the selectedTrain. You just need to transform it to the locale timeZone.
- add a new utils function in utils/date.ts
```
export const isoUtcStringToLocaleDateString = (isoUtcString: string) => {
  return dayjs(isoUtcString).local().format();
};
```
- give it to TimesStops
```suggestion
      startTime={isoUtcStringToLocaleDateString(selectedTrainSchedule.start_time)}
```

anisometropie replied:
done

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -0,0 +1,117 @@
+        if (pathStepKey in pathStepsByUic && nextOpPoint) {
+          const pathStepId = pathStepsByUic[pathStepKey].id || '';
+          const schedule = scheduleByAt[pathStepId];
+          const scheduleData = formatScheduleData(schedule, startTime);
```
```suggestion
          const scheduleData = formatScheduleData(schedule, selectedTrainSchedule.start_time);
```

anisometropie replied:
done

flomonster commented:
Column titles in English doesn't fit:

![image](https://github.com/user-attachments/assets/20ca2286-f0e7-47db-9ef6-144b1bb7e277)

clarani reviewed (approved):
LGTM :white_check_mark:
